docs/proposals/config-management-plugin-v2.md

```diff
@@ -0,0 +1,304 @@
+
+### Risks and Mitigations
+
+One issue is that currently when repositories are cloned, the repo is cloned using the same UID of the repo-server user, 
+and so all repository files are created using that UID. This means that a command which executes in the git repository path, 
+could traverse upwards and see/write files which are outside of the repository tree.
+
+One proposal to prevent out-of-tree access to files, is that each git repository could be cloned with unique UIDs, 
+different from the repo-server’s UID. When the cmp-server executes the tooling command to generate manifests, 
+the command could be executed using the UID of the git repository files. e.g.:
+
+```
+cmd := exec.Command(command, args...)
+cmd.SysProcAttr = &syscall.SysProcAttr{}
+cmd.SysProcAttr.Credential = &syscall.Credential{Uid: uid, Gid: gid}
+```
+
+This would ensure that the command could not read or write anything out-of-tree from the repository directory.
+
+### Upgrade / Downgrade Strategy
+
+The argocd-repo-server manifest will change in order to populate the argocd-cmp-server binary inside the plugin container 
+via an init container, which will pre-populate a volume shared between plugins and the repo-server.
+
+```bash
+# An init container will copy the argocd static binary into the shared volume
+# so that the CMP server can become the entrypoint
+initContainers:
+- command:
+  - cp
+  - -n
+  - /usr/local/bin/argocd
+  - /var/run/argocd/argocd-cmp-server
+  image: quay.io/argoproj/argocd:latest
+  name: copyutil
+  volumeMounts:
+  - mountPath: /var/run/argocd
+    name: var-files
+ 
+# var-files is a shared volume between repo-server and cmp-server which holds:
+# 1) socket files that repo-server uses to communicate to each plugin
+# 2) git repositories cloned by repo-server
+volumes:
+- emptyDir: {}
+  name: var-files
+```
+  
+After upgrading to CMP v2, an Argo CD operator will have to make following changes,
+
+* In order to install a plugin, an Argo CD operator will simply have to patch argocd-repo-server 
+to run config management plugin container as a sidecar, with argocd-cmp-server as it’s entrypoint:
+
+```bash
+# A plugin is a container image which runs as a sidecar, with the execution environment
+# necessary to render manifests. To install a plugin, 
+containers:
+- name: cdk8s
+  command: [/var/run/argocd/argocd-cmp-server]
+  image: cdk8s/cdk8s:latest
+  volumeMounts:
+  - mountPath: /var/run/argocd
+    name: var-files
+```
+
+* Plugins will be configured via a ConfigManagementPlugin manifest located inside the plugin container, placed at a 
+well-known location (e.g. /plugin.yaml). Argo CD is agnostic to the mechanism of how the plugin.yaml would be placed, 
+but various options can be used on how to place this file, including: 
+    - Baking the file into the plugin image as part of docker build
+    - Volume mapping the file through a configmap.
+(For more details please refer to [implementation details](#configuration))
+
+## Drawbacks
```

Comment:
You may mention that there aren't any major drawbacks and that the advantages supersede the minor learning curve of the new way of managing plugins.

However, 

* More yaml ;)
* Operators need to be aware of the modified kubernetes manifests in the subsequent version.
* The format of the CMP manifest is a new "contract" that would need to adhere the usual Argo CD compatibility promises in future.

Reply:
Thank you for listing these.